### PR TITLE
Intrepid2: Initial implementation of team-level getValues

### PR DESCRIPTION
**Pull request by mperego**
@trilinos/intrepid2 

## Motivation
In most cases, we want to evaluate the reference basis functions at a set of reference points (e.g. quadrature points) that is the same for all the cells in a workset. In this case one evaluates the basis functions at these points on the reference cell, using the `getValues` function, and then transforms these basis functions to physical cells.
In some cases, however, the evaluation points can be different from cell to cell (e.g., in particle in cell methods, we can have different particles in different cells where we want to evaluate the basis functions). For this use-case, we implement a team-level `getValues` function, that can be called by the user in a parallel for using a `TeamPolicy`. 
This PR:
- Implements team-level getValues for classic Lagrangian basis functions.
- Modifies/adds tests to compare the team-level getValues with host getValues
- Modifies impelementation of JacobiPolynomial to reduce FAD temporaries, which could cause running out of space for device-side dynamic allocations

### CODE REVIEW

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp [hunk 1]
```diff
@@ -379,6 +379,54 @@ using HostBasisPtr = BasisPtr<typename Kokkos::HostSpace::device_type, OutputTyp
...
+        \param  perTeamSpaceSize  [out] - size of the scratch space needed per team
+        \param  perThreadeSize    [out] - size of the scratch space beeded per thread
+    */
+    virtual
+    void getScratchSpaceSize(       ordinal_type& perTeamSpaceSize,
+                                    ordinal_type& perThreadSpaceSize,
+                                const PointViewType inputPoints,
+                                const EOperator operatorType = OPERATOR_VALUE) const {
+      INTREPID2_TEST_FOR_EXCEPTION_DEVICE_SAFE( true, std::logic_error,
+                                    ">>> ERROR (Basis::getValuesScratchSpace): this method (FEM) is not supported or should be overridden accordingly by derived classes.");
```
Comment (CamelliaDPG) [inline packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp:400]: What is the "(FEM)" supposed to convey here?  I think the message is clearer without it, but maybe there is something else you wanted to communicate.

**Comment by mperego** [comment]
@CamelliaDPG @rppawlo , I encounter this error on CUDA with `HCURL_HEX_In` basis, `test_02`, when using dynamic Fad type and enabling `Sacado_ENABLE_HIERARCHICAL_DFAD`:
```
14: :0: : block: [3,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
14: :0: : block: [3,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
14: :0: : block: [3,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
14: :0: : block: [0,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
14: :0: : block: [0,0,0], thread: [0,2,0] Assertion `Allocation failed.` failed.
14: :0: : block: [0,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
14: :0: : block: [0,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
14: :0: : block: [2,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
14: :0: : block: [2,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
14: :0: : block: [2,0,0], thread: [0,2,0] Assertion `Allocation failed.` failed.
14: :0: : block: [2,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
14: :0: : block: [1,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
14: :0: : block: [1,0,0], thread: [0,2,0] Assertion `Allocation failed.` failed.
14: :0: : block: [1,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
14: :0: : block: [1,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
14: :0: : block: [4,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
14: :0: : block: [4,0,0], thread: [0,2,0] Assertion `Allocation failed.` failed.
14: :0: : block: [4,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
14: :0: : block: [4,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
14: (ptr->cuda_stream_synchronize_wrapper(stream)) error( cudaErrorAssert): device-side assert triggered /home/mperego/weaver/trilinos/trilinos-repo/packages/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:166
14: Backtrace:
14: [0x20003600a644] Kokkos::Impl::save_stacktrace()
14: [0x200035fd8598] Kokkos::Impl::host_abort(char const*)
14: [0x200036014b14] Kokkos::Impl::cuda_internal_error_abort(cudaError, char const*, char const*, int)
14: [0x200036016654] Kokkos::Impl::cuda_stream_synchronize(CUstream_st*, Kokkos::Impl::CudaInternal const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
14: [0x2000360166f4] Kokkos::Impl::CudaInternal::fence(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
14: [0x2000360168bc] Kokkos::Cuda::fence(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
14: [0x100d5b20] 
14: [0x101ebb2c] 
14: [0x1001ee6c] 
14: [0x200036949f5c] 
14: [0x20003694a0f4] __libc_start_main
```
Any clue of what could be causing it? It does happen only for high-order (order 9), however, I don't think it is a out of memory issue, as the `HCURL_TET_In` `test_02` uses much more scratch memory and it runs fine.

**Comment by CamelliaDPG** [comment]
> @CamelliaDPG @rppawlo , I encounter this error on CUDA with `HCURL_HEX_In` basis, `test_02`, when using dynamic Fad type and enabling `Sacado_ENABLE_HIERARCHICAL_DFAD`:
> 
> ```
> 14: :0: : block: [3,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [3,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [3,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [0,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [0,0,0], thread: [0,2,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [0,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [0,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [2,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [2,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [2,0,0], thread: [0,2,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [2,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [1,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [1,0,0], thread: [0,2,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [1,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [1,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [4,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [4,0,0], thread: [0,2,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [4,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
> 14: :0: : block: [4,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
> 14: (ptr->cuda_stream_synchronize_wrapper(stream)) error( cudaErrorAssert): device-side assert triggered /home/mperego/weaver/trilinos/trilinos-repo/packages/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:166
> 14: Backtrace:
> 14: [0x20003600a644] Kokkos::Impl::save_stacktrace()
> 14: [0x200035fd8598] Kokkos::Impl::host_abort(char const*)
> 14: [0x200036014b14] Kokkos::Impl::cuda_internal_error_abort(cudaError, char const*, char const*, int)
> 14: [0x200036016654] Kokkos::Impl::cuda_stream_synchronize(CUstream_st*, Kokkos::Impl::CudaInternal const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
> 14: [0x2000360166f4] Kokkos::Impl::CudaInternal::fence(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
> 14: [0x2000360168bc] Kokkos::Cuda::fence(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
> 14: [0x100d5b20] 
> 14: [0x101ebb2c] 
> 14: [0x1001ee6c] 
> 14: [0x200036949f5c] 
> 14: [0x20003694a0f4] __libc_start_main
> ```
> 
> Any clue of what could be causing it? It does happen only for high-order (order 9), however, I don't think it is a out of memory issue, as the `HCURL_TET_In` `test_02` uses much more scratch memory and it runs fine.

Sorry I missed the question last week -- I'm not sure; it does sound to me like it is running out of memory.  Do you know which kernel it's in when it fails?  Happy to look at the particular code more carefully to see what jumps out.

**Comment by mperego** [comment]
> > @CamelliaDPG @rppawlo , I encounter this error on CUDA with `HCURL_HEX_In` basis, `test_02`, when using dynamic Fad type and enabling `Sacado_ENABLE_HIERARCHICAL_DFAD`:
> > ```
> > 14: :0: : block: [3,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [3,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [3,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [0,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [0,0,0], thread: [0,2,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [0,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [0,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [2,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [2,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [2,0,0], thread: [0,2,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [2,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [1,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [1,0,0], thread: [0,2,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [1,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [1,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [4,0,0], thread: [0,1,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [4,0,0], thread: [0,2,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [4,0,0], thread: [0,3,0] Assertion `Allocation failed.` failed.
> > 14: :0: : block: [4,0,0], thread: [0,0,0] Assertion `Allocation failed.` failed.
> > 14: (ptr->cuda_stream_synchronize_wrapper(stream)) error( cudaErrorAssert): device-side assert triggered /home/mperego/weaver/trilinos/trilinos-repo/packages/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:166
> > 14: Backtrace:
> > 14: [0x20003600a644] Kokkos::Impl::save_stacktrace()
> > 14: [0x200035fd8598] Kokkos::Impl::host_abort(char const*)
> > 14: [0x200036014b14] Kokkos::Impl::cuda_internal_error_abort(cudaError, char const*, char const*, int)
> > 14: [0x200036016654] Kokkos::Impl::cuda_stream_synchronize(CUstream_st*, Kokkos::Impl::CudaInternal const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
> > 14: [0x2000360166f4] Kokkos::Impl::CudaInternal::fence(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
> > 14: [0x2000360168bc] Kokkos::Cuda::fence(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
> > 14: [0x100d5b20] 
> > 14: [0x101ebb2c] 
> > 14: [0x1001ee6c] 
> > 14: [0x200036949f5c] 
> > 14: [0x20003694a0f4] __libc_start_main
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Any clue of what could be causing it? It does happen only for high-order (order 9), however, I don't think it is a out of memory issue, as the `HCURL_TET_In` `test_02` uses much more scratch memory and it runs fine.
> 
> Sorry I missed the question last week -- I'm not sure; it does sound to me like it is running out of memory. Do you know which kernel it's in when it fails? Happy to look at the particular code more carefully to see what jumps out.

@etphipp helped me figure this out. It's related to the creation of dFAD temporaries.  Sacado is erroring out when there is not enough memory to create a new dFAD type. I've being fixing this by reducing the number of temporaries needed.

**Comment by mperego** [comment]
@CamelliaDPG I think this is ready to be reviewed. Thanks!

**Comment by mperego** [comment]
@CamelliaDPG , All tests pass now except for the Polylib test which times out on one debug PR build.
I had to refactor the `JacobiPolynomial` function to reduce the number of FAD temporaries and comparing the timings with those from your PR on increasing` MaxOrder` it got a bit slower, in some cases up to 20% slower. There is only one PR build where the new implementation is faster. 
What do you think? I can go back to the previous design, but I would have to pass an extra view to `JacobiPolynomial` to be used as temporary storage. 
If you think the slowdown is acceptable, then we have to find a way to make the test shorter (which is probably a good idea anyway). Do you have any suggestion? The most obvious thing is to reduce the max number of integration/interpolation points used in the test.

File: packages/intrepid2/unit-test/Shared/Polylib/test_01.hpp [hunk 2]
```diff
@@ -316,6 +320,7 @@ namespace Intrepid2 {
                         ", np = " << np << ", n = " << n << "  difference " << sum << "\n";
                     }
                   }
+                  if (np == npUpperStep1) npStep = 5;
```
Comment (mperego) [inline packages/intrepid2/unit-test/Shared/Polylib/test_01.hpp:323]: I like this.

**Review by CamelliaDPG — APPROVED** [review]
Looks good to me, @mperego!

File: packages/intrepid2/src/Shared/Intrepid2_Utils.hpp [hunk 3]
```diff
@@ -805,6 +805,32 @@ namespace Intrepid2 {
...
+  };
+
+  /// Function for creating a vtable on device (requires copy ctor for
+  /// derived object). Allocates device memory and must be called from
+  /// host.
+  template<typename Device,typename Derived>
+  std::unique_ptr<Derived,DeviceDeleter<Device>>
+  copy_virtual_class_to_device(const Derived& host_source)
+  {
+    auto* p = static_cast<Derived*>(Kokkos::kokkos_malloc<typename Device::memory_space>(sizeof(Derived)));
```
[the remark this answers is not in the document]
Comment (mperego) [reply packages/intrepid2/src/Shared/Intrepid2_Utils.hpp:828]: It's used in the new tests, e.g.
https://github.com/mperego/Trilinos/blob/intrepid2_parallel_getValues/packages/intrepid2/unit-test/Discretization/Basis/HGRAD_HEX_Cn_FEM/test_02.hpp#L80

**Comment by mperego** [comment]
Thanks @CamelliaDPG!
We need to sign off the commits now (see the failing Per-Commit Checks) . 
Do you mind if I squash your commit into mine?
Alternatively you ave to amend your commit and force push it.
`git commit -s --amend 748d94d84`

**Comment by CamelliaDPG** [comment]
> Thanks @CamelliaDPG! We need to sign off the commits now (see the failing Per-Commit Checks) . Do you mind if I squash your commit into mine? Alternatively you ave to amend your commit and force push it. `git commit -s --amend 748d94d84`

I don't mind if you squash my commit into yours.  Thanks!

**Comment by mperego** [comment]
@CamelliaDPG I squashed the commits. Would you mind re-approving?

File: packages/intrepid2/src/Shared/Intrepid2_Utils.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (CamelliaDPG) [reply packages/intrepid2/src/Shared/Intrepid2_Utils.hpp:828]: Ah, got it — makes good sense.  I think I must have been searching under the `src` directory, not looking in the tests.

**Comment by CamelliaDPG** [comment]
> @CamelliaDPG I squashed the commits. Would you mind re-approving?

I think I don't need to; I think because it was just a squash my prior approval persists.

**Pull request merged**